source/dofs/dof_vector.cc

```diff
@@ -0,0 +1,91 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2009 - 2015 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/dofs/dof_vector.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/hp/dof_handler.h>
+#include <deal.II/base/mg_level_object.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/block_vector.h>
+#include <deal.II/lac/parallel_vector.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace
+{
+  template <class DH, class VECTOR>
+  void reinit_dof_vector (const DH &, VECTOR &)
+  {
+    Assert(false, ExcNotImplemented());
+  }
+
+  template <int dim, int spacedim, typename number>
+  void reinit_dof_vector (const DoFHandler<dim, spacedim> &dh,
+                          Vector<number> &v)
+  {
+    v.reinit(dh.n_dofs());
+  }
+
+  template <int dim, int spacedim, typename number>
+  void reinit_dof_vector (const DoFHandler<dim, spacedim> &dh,
+                          MGLevelObject<Vector<number> > &v)
+  {
+    const unsigned int nlev = dh.get_tria().n_levels();
+    v.resize(0, nlev-1);
+    for (unsigned int l=0; l<nlev; ++l)
+      v[l].reinit(dh.n_dofs(l));
+  }
+
+  template <int dim, int spacedim, typename number>
+  void reinit_dof_vector (const DoFHandler<dim, spacedim> &dh,
+                          BlockVector<number> &v)
+  {
+    v.reinit(dh.block_info().global());
```

Comment:
This assumes a very specific data layout that may or may not be what the user has in mind.